### PR TITLE
vimPlugins: backoff on timeout in update.py

### PR DESCRIPTION
**Pull request by timokau**
###### Motivation for this change

Updating vim-plugins recently started timing out regularly for me. It
may have to do with an ISP switch on my side, but I don't think that
should cause timeouts. I guess I'm probably not the only one
experiencing this, so in this comment I introduce exponential backoff.
Every request will be retried up to 3 times (3 seconds, 6 seconds, 12
seconds delay).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

CC @Mic92 @rvolosatovs

### CODE REVIEW

File: pkgs/misc/vim-plugins/update.py [hunk 1]
```diff
@@ -24,12 +24,59 @@
...
+ATOM_LINK = "{http://www.w3.org/2005/Atom}link" #"
+ATOM_UPDATED = "{http://www.w3.org/2005/Atom}updated" #"
 
 ROOT = Path(__file__).parent
 
+import time
+from functools import wraps
+
+
+def retry(ExceptionToCheck, tries=4, delay=3, backoff=2, logger=None):
```
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py]: Can you add type hints here as well?

File: pkgs/misc/vim-plugins/update.py [hunk 2]
```diff
@@ -24,12 +24,59 @@
...
+    :type tries: int
+    :param delay: initial delay between retries in seconds
+    :type delay: int
+    :param backoff: backoff multiplier e.g. value of 2 will double the delay
+        each retry
+    :type backoff: int
+    :param logger: logger to use. If None, print
+    :type logger: logging.Logger instance
+    """
+    def deco_retry(f):
```
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py]: And here.

File: pkgs/misc/vim-plugins/update.py [hunk 3]
```diff
@@ -24,12 +24,59 @@
...
+
+        @wraps(f)
+        def f_retry(*args, **kwargs):
```
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py]:
```suggestion
        def f_retry(*args: Any, **kwargs: Any) -> Any:
```

File: pkgs/misc/vim-plugins/update.py (hunk 2 above)
Comment (Mic92) [reply pkgs/misc/vim-plugins/update.py]:
```suggestion
    def deco_retry(f: Callable) -> Callable:
```

File: pkgs/misc/vim-plugins/update.py (hunk 1 above)
Comment (Mic92) [reply pkgs/misc/vim-plugins/update.py]:
```suggestion
def retry(ExceptionToCheck: Exception, tries: int = 4, delay : int = 3, backoff: int = 2):
```

File: pkgs/misc/vim-plugins/update.py [hunk 4]
```diff
@@ -24,12 +24,59 @@
...
+        exceptions to check
+    :type ExceptionToCheck: Exception or tuple
+    :param tries: number of times to try (not retry) before giving up
+    :type tries: int
+    :param delay: initial delay between retries in seconds
+    :type delay: int
+    :param backoff: backoff multiplier e.g. value of 2 will double the delay
+        each retry
+    :type backoff: int
+    :param logger: logger to use. If None, print
```
Comment (Mic92) [inline pkgs/misc/vim-plugins/update.py]: The logger is unused and can be removed.

File: pkgs/misc/vim-plugins/update.py (hunk 1 above)
Comment (Mic92) [reply pkgs/misc/vim-plugins/update.py]: All the type hints are untested, so you may want to check with mypy.

**Comment by timokau** [comment]
Addressed your comments. Also fixed unrelated formatting and typing issues.

File: pkgs/misc/vim-plugins/update.py (hunk 1 above)
Comment (timokau) [reply pkgs/misc/vim-plugins/update.py]: Unfortunately `Exception` didn't work, maybe python special cases that? mypy complained:

```
update.py:58: error: Exception type must be derived from BaseException
update.py:81: error: Argument 1 to "retry" has incompatible type "Type[URLError]"; expected "Exception"
update.py:94: error: Argument 1 to "retry" has incompatible type "Type[URLError]"; expected "Exception"
```

So I used "Any" instead (which also respects that you could theoretically pass in a tuple of Exceptions).

**Comment by Mic92** [comment]
Unfortunately we now have a merging conflict.

**Comment by timokau** [comment]
Rebased.

**Comment by timokau** [comment]
@Mic92 please have another look :)

**Pull request merged**